### PR TITLE
MicroMeter metrics implementation

### PR DESCRIPTION
**Pull request by alexanderst**

### CODE REVIEW

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineGauge.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        mmGauge =  io.micrometer.core.instrument.Gauge.builder(id, value, new ToDoubleFunction<Double>() {
+            @Override
+            public double applyAsDouble(Double value) {
+                return value;
+            }
+        }).register(MicrometerRegistry.getRegistry());
+    }
+
+    public PipelineGauge(String id, List<?> list) {
+        mmGauge = io.micrometer.core.instrument.Gauge.builder(id, list, List::size).register(new SimpleMeterRegistry());
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineGauge.java:39]: `.register(MicrometerRegistry.getRegistry())`

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineTimer.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ */
+package ai.konduit.serving.pipeline.api.context;
+
+import ai.konduit.serving.pipeline.registry.MicrometerRegistry;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+public class PipelineTimer implements Timer {
+    io.micrometer.core.instrument.Timer mmTimer;
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineTimer.java:26]: private access modifier

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineTimer.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @Override
+    public void record(long duration, TimeUnit timeUnit) {
+        mmTimer.record(duration, timeUnit);
+    }
+
+    @Override
+    public long stop() {
+        return io.micrometer.core.instrument.Timer.start().stop(mmTimer);
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineTimer.java:44]: This start().stop(...) won't return the correct time.
Looks like instructions were bad here: https://github.com/KonduitAI/konduit-serving/issues/326

The workflow for the user would be:
```
Timer t = ctx.metrics().timer("timer");
Sample s = t.start();   //Records start time
t.stop(s);
```

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/Timer.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+public interface Timer {
+
+    void record(Duration duration);
+
+    void record(long duration, TimeUnit timeUnit);
+
+    long stop();
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/Timer.java:29]: As per previous comment:
```
Sample start();
long stop(Sample);
```

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/registry/MicrometerRegistry.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ *  * SPDX-License-Identifier: Apache-2.0
+ *  *****************************************************************************
+ */
+package ai.konduit.serving.pipeline.registry;
+
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+
+public class MicrometerRegistry {
+    public static io.micrometer.core.instrument.MeterRegistry getRegistry() {
+        return io.micrometer.core.instrument.Metrics.globalRegistry.add(new SimpleMeterRegistry());
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/registry/MicrometerRegistry.java:24]: We shouldn't be adding a new registry on every call.

```
public class MicrometerRegistry {
    static {
        io.micrometer.core.instrument.Metrics.globalRegistry.add(new SimpleMeterRegistry());
    }
    public static io.micrometer.core.instrument.MeterRegistry getRegistry() {
        return io.micrometer.core.instrument.Metrics.globalRegistry;
    }
}
```

Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/registry/MicrometerRegistry.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/Timer.java (hunk 4 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/Timer.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineTimer.java (hunk 3 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineTimer.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineTimer.java (hunk 2 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineTimer.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineGauge.java (hunk 1 above)
Comment (alexanderst) [reply konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/PipelineGauge.java]: Fixed.

File: konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/Timer.java [hunk 6]
```diff
@@ -22,9 +22,13 @@
 
 public interface Timer {
 
+    public interface Sample {
+        long stop(Timer timer);
```
Comment (AlexDBlack) [inline konduit-serving-pipeline/src/main/java/ai/konduit/serving/pipeline/api/context/Timer.java:26]: This should be `stop(Sample sample)`

**Pull request merged**